File: CirQueue.c
#include <stdlib.h>
#include <stdio.h>

/**
 * p.75: 循环队列
 */
#define QUEUESIZE 20

typedef char DataType;

typedef struct {
    DataType data[QUEUESIZE];
    int front, rear;
} CirQueue;

/**
 * p.75: 置空队列
 */
void InitQueue(CirQueue *Q) {
    Q->front = Q->rear = 0;
}

/**
 * p.76: 判队空
 */
int QueueEmpty(CirQueue *Q) {
    return Q->rear == Q->front;
}

/**
 * p.76: 判队满
 */
int QueueFull(CirQueue *Q) {
    return (Q->rear + 1) % QUEUESIZE == Q->front;
}

/**
 * p.76: 入队列
 */
void EnQueue(CirQueue *Q, DataType x) {
    if (QueueFull(Q)) {
        printf("Queue overflow\n");
    } else {
        Q->data[Q->rear] = x;
        Q->rear = (Q->rear + 1) % QUEUESIZE; // 使用其余运算实现循环意义下的加 1
    }
}

/**
 * p.76: 取队头元素
 */
DataType GetFront(CirQueue *Q) {
    if (QueueEmpty(Q)) {
        printf("Queue empty\n");
        exit(0);
    } else {
        return Q->data[Q->front];
    }
}

/**
 * p.76: 出队列
 */
DataType DeQueue(CirQueue *Q) {
    DataType x;
    if (QueueEmpty(Q)) {
        printf("Queue empty\n");
        exit(0);
    }
    x = Q->data[Q->front];
    Q->front = (Q->front + 1) % QUEUESIZE;
    return x;
}

File: go/cirqueue/cirqueue.go
package cirqueue

import "errors"

// Q is an circular queue.
type Q[T any] struct {
	data        []T
	len         int
	front, rear int
}

var (
	errFull = errors.New("queue is full")
	errEmpty = errors.New("queue is empty")
)

// New create a Q which can hold n elements.
func New[T any](n uint) *Q[T] {
	// NOTE: to distinguish empty/full check, rear always point to a vacant
	// position, so we need n+1 positions.
	return &Q[T]{data: make([]T, n+1), len: int(n+1)}
}

// IsEmpty check whether q is empty.
func (q *Q[T]) IsEmpty() bool {
	return q.front == q.rear
}

// IsFull check whether q is full.
func (q *Q[T]) IsFull() bool {
	return ((q.rear+1) % q.len) == q.front
}

// Enqueue append new element of data x into q.
func (q *Q[T]) Enqueue(x T) error {
	if q.IsFull() {
		return errFull
	}
	q.data[q.rear] = x
	q.rear = (q.rear + 1) % q.len
	return nil
}

// Dequeue return next element in q, FIFO style.
func (q *Q[T]) Dequeue() (T, error) {
	var x T
	if q.IsEmpty() {
		return x, errEmpty
	}
	x = q.data[q.front]
	q.front = (q.front+1) % q.len
	return x, nil
}

// Front return next element in q, FIFO style, without removing it.
func (q *Q[T]) Front() (T, error) {
	var x T
	if q.IsEmpty() {
		return x, errEmpty
	}
	return q.data[q.front], nil
}